src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
+    for (ValueID dictionary_index = ValueID{0}; dictionary_index < dictionary_size; ++dictionary_index) {
+      if (value_by_value_id(dictionary_index) >= value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(static_cast<T>(value)); }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    auto dictionary_size = _dictionary->size();
+    for (ValueID dictionary_index = ValueID{0}; dictionary_index < dictionary_size; ++dictionary_index) {
+      if (value_by_value_id(dictionary_index) > value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(static_cast<T>(value)); }
 
-  // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  // return the number of _dictionary (dictionary entries)
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    auto attributeVecMem = _attribute_vector->width() * _attribute_vector->size();
+    auto dictionaryVecMem = _dictionary->size() * sizeof(T);
+    return static_cast<size_t>(attributeVecMem + dictionaryVecMem);
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  void _build_compressed_dictionary(const std::vector<T>& values) {
+    auto raw_dictionary = std::move(values);
+    std::vector<T> raw_values = raw_dictionary;
+
+    std::sort(raw_dictionary.begin(), raw_dictionary.end());
+    raw_dictionary.erase(std::unique(raw_dictionary.begin(), raw_dictionary.end()), raw_dictionary.end());
+
+    _dictionary = std::make_shared<std::vector<T>>(raw_dictionary);
+
+    _build_attribute_vector(raw_dictionary, raw_values);
+  }
+
+  void _build_attribute_vector(std::vector<T>& raw_dictionary, std::vector<T>& all_values) {
+    std::vector<uint32_t> raw_vector{};
+    raw_vector.reserve(all_values.size());
+    size_t size = all_values.size();
+
+    const int required_width = std::ceil(std::log2(size));
+
+    if (required_width <= 8) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(size);
```

Comment:
If `size` is `256` you would choose an 8bit AttributeVector which would leave no space for `INVALID_VALUE_ID`.